### ISSUE TITLE
getHistory requests data "Today - retentionPeriod" when DB did not exist and errors

### ISSUE THREAD

**Issue by wiki-io** [ioBroker/ioBroker.influxdb#222]
Hello,
I am strugling with an issue of the influxdb2 adapter since a while.
First I thought its root cause was Jarvis, but I think its an issue of the influxdb2 adapter.

Forum-Post (german): https://forum.iobroker.net/post/747718

![grafik](https://user-images.githubusercontent.com/99530859/153704649-82c3c3b0-87bd-497c-a0d5-7780b067eea4.png)

When the Charting is requesting the data a a given timeframe (x days/weeks back from today), the influx adapter suddenly inserts a query of data for "today - retention period". 
As the DB exists just a few month the query fails.
All subsequent queries fail as well and all chartings get unusable.

`
2022-01-22 17:35:45.222  - [34mdebug[39m: influxdb.0 (1908) History-queries to execute: 
from(bucket: "iobroker")  |> range(start: 2021-07-09T16:35:39.880Z, stop: 2022-01-08T16:35:39.880Z)     |> filter(fn: (r) => r["_measurement"] == "tado.0.667342.Rooms.4.sensorDataPoints.humidity.percentage") 	|> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value") |> sort(columns: ["_time"], desc: true)  	|> group() 	|> limit(n: 1),
from(bucket: "iobroker")  |> range(start: 2022-01-08T16:35:39.880Z, stop: 2022-01-22T16:35:39.880Z) 	|> filter(fn: (r) => r["_measurement"] == "tado.0.667342.Rooms.4.sensorDataPoints.humidity.percentage") 	|> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value") 											|> group() 	|> limit(n: 9999999),
from(bucket: "iobroker")  |> range(start: 2022-01-22T16:35:39.880Z)                 					|> filter(fn: (r) => r["_measurement"] == "tado.0.667342.Rooms.4.sensorDataPoints.humidity.percentage")  	|> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value") |> sort(columns: ["_time"], desc: false) 	|> group() 	|> limit(n: 1)

Execution of Query - Part I:
2022-01-22 17:35:45.222  - [34mdebug[39m: influxdb.0 (1908) Query to execute: 
from(bucket: "iobroker")  |> range(start: 2021-07-09T16:35:39.880Z, stop: 2022-01-08T16:35:39.880Z)     |> filter(fn: (r) => r["_measurement"] == "tado.0.667342.Rooms.4.sensorDataPoints.humidity.percentage")                 |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value")                |> sort(columns: ["_time"], desc: true)                 |> group()                 |> limit(n: 1)

Execution of Query - Part I - klar DB existierte ja noch nicht.
2022-01-22 17:35:56.911  - [33mwarn[39m: influxdb.0 (1908) Error in query "
from(bucket: "iobroker")  |> range(start: 2021-07-09T16:35:39.880Z, stop: 2022-01-08T16:35:39.880Z)     |> filter(fn: (r) => r["_measurement"] == "tado.0.667342.Rooms.4.sensorDataPoints.humidity.percentage")                 |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value")                |> sort(columns: ["_time"], desc: true)                 |> group()                 |> limit(n: 1)": RequestTimedOutError: Request timed out

Execution of Query - Part II:
2022-01-22 17:35:56.911  - [34mdebug[39m: influxdb.0 (1908) Query to execute: 
from(bucket: "iobroker")  |> range(start: 2022-01-08T16:35:39.880Z, stop: 2022-01-22T16:35:39.880Z) 	|> filter(fn: (r) => r["_measurement"] == "tado.0.667342.Rooms.4.sensorDataPoints.humidity.percentage") |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value") |> group() |> limit(n: 9999999)

Execution of Query - Part III:
2022-01-22 17:35:58.792  - [34mdebug[39m: influxdb.0 (1908) Query to execute: from(bucket: "iobroker")                 |> range(start: 2022-01-22T16:35:39.880Z)                 |> filter(fn: (r) => r["_measurement"] == "tado.0.667342.Rooms.4.sensorDataPoints.humidity.percentage")                 |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value")                |> sort(columns: ["_time"], desc: false)                 |> group()                 |> limit(n: 1)

2022-01-22 17:36:00.454  - [33mwarn[39m: jarvis.0 (1876) [object Object]

Execution of Query - Part II - Receive Data:
2022-01-22 17:35:59.614  - [34mdebug[39m: influxdb.0 (1908) Parsing retrieved rows:[
[],
[{"result":"_result","table":0,"_start":"2022-01-08T16:35:39.88Z","_stop":"2022-01-22T16:35:39.88Z","_time":"2022-01-08T16:35:50.919Z","_measurement":"tado.0.667342.Rooms.4.sensorDataPoints.humidity.percentage","ack":true,"from":"system.adapter.influxdb.0","q":0,"value":44.5,"time":"2022-01-08T16:35:50.919Z"},
{"result":"_result","table":0,"_start":"2022-01-08T16:35:39.88Z","_stop":"2022-01-22T16:35:39.88Z","_time":"2022-01-08T16:36:50.92Z","_measurement":"tado.0.667342.Rooms.4.sensorDataPoints.humidity.percentage","ack":true,"from":"system.adapter.influxdb.0","q":0,"value":44.5,"time":"2022-01-08T16:36:50.92Z"},
...
{"result":"_result","table":0,"_start":"2022-01-08T16:35:39.88Z","_stop":"2022-01-22T16:35:39.88Z","_time":"2022-01-22T16:29:11.726Z","_measurement":"tado.0.667342.Rooms.4.sensorDataPoints.humidity.percentage","ack":true,"from":"system.adapter.tado.0","q":0,"value":39.7,"time":"2022-01-22T16:29:11.726Z"}],[]]

Data to Jarvis:
2022-01-22 17:35:59.936  - [34mdebug[39m: influxdb.0 (1908) sendTo "getHistory" to system.adapter.jarvis.0 from system.adapter.influxdb.0
`

I had a look at the code and I think the function that is triggering the bevaiour is maybe:
`
 //Workaround to detect if measurement is of type bool (to skip non-sensical aggregation options)
    //There seems to be no officially supported way to detect this, so we check it by forcing a type-conflict
    const booleanTypeCheckQuery = '\
        from(bucket: "' + adapter.config.dbname + '")\
        |> range(' + ((options.start) ? 'start: ' + new Date(options.start).toISOString() + ', ' : 'start: -' + adapter.config.retention +'ms, ') + 'stop: ' + new Date(options.end).toISOString() + ')\
        |> filter(fn: (r) => r["_measurement"] == "' + options.id + '" and contains(value: r._value, set: [true, false]))\
    ';
`

Hope this can be improved.

best, wiki

**Comment by Excodibur** [comment]
T.b.h. I don't know much about Jarvis and how it queries data via influxdb, but I am relatively sure that setting the _start_-date as part of the _range()_ function to a date prior to DB creation should not cause issues like you observed.

I tested it on my local Influx 2 test db, and Influx does not have any problems, even if set the _start_ date to e.g. the year 1900. 

With Influx 2 all Flux-queries require a time range to be set, and part of that _range_-function is the _start_ parameter that is mandatory. As there is no simple way to set _start_ to "when the db was first created", the only option we found is to estimate the oldest date a potential record could have, and that is the current date - max retention time (after which records are dropped). But as said before, even if we would go back further InfluxDB should be able to handle this.

I think the problem here could be, that your read-request (as the messages states) times out, likely because it took too long to gather all the data on server side.

The default timeout set by the client library we use seems to be 10 seconds:
https://github.com/influxdata/influxdb-client-js/blob/2fff485313d85d14b79daa9552e93461e73c0a65/packages/core/src/options.ts#L12-L16

Which is roughly the passed time that can be seen in your logs from request to error.

We could try to make it configurable somewhere [here](https://github.com/ioBroker/ioBroker.influxdb/blob/829c2b909151846f5ce910e24d0ebe9225b5177e/lib/DatabaseInfluxDB2x.js#L50-L53) to give the adapter more time to fetch the result.

Let me check it.
